source/common/http/async_client_impl.cc

```diff
@@ -143,16 +143,30 @@ void AsyncStreamImpl::sendTrailers(HeaderMap& trailers) {
 
 void AsyncStreamImpl::closeLocal(bool end_stream) {
   ASSERT(!(local_closed_ && end_stream));
+  // Due to the fact that send calls can synchronously result in stream closure, it's possible for
+  // this to be called after the local side is already closed.
+  if (local_closed_) {
+    return;
+  }
 
-  local_closed_ |= end_stream;
+  local_closed_ = end_stream;
   if (complete()) {
+    stream_callbacks_.onComplete();
     cleanup();
   }
 }
 
 void AsyncStreamImpl::closeRemote(bool end_stream) {
-  remote_closed_ |= end_stream;
+  // Due to the fact that callbacks can synchronously result in stream closure, it's possible for
```

Comment:
Again a bit hazy on how this happens. Can you provide more info?

Reply:
In this case, it's possible when a callback synchronously calls `stream->reset`. There's a unit test that specifically covers this.

Reply:
But if closes only happen in non-reset cases, why are we calling closeRemote() in the case of reset?

/wait-any

Reply:
Let's assume the case covered by the unit test.
We're in `encodeHeaders`: https://github.com/envoyproxy/envoy/blob/ef543d8fcea4d1c0ce85e3ff8d16fb3928436ed6/source/common/http/async_client_impl.cc#L89

`stream_callbacks.onHeaders` is called and synchronously resets the stream.  `reset` calls `resetStream` which calls `cleanup`: https://github.com/envoyproxy/envoy/blob/ef543d8fcea4d1c0ce85e3ff8d16fb3928436ed6/source/common/http/async_client_impl.cc#L190

`cleanup` sets local and remote to be closed: https://github.com/envoyproxy/envoy/blob/ef543d8fcea4d1c0ce85e3ff8d16fb3928436ed6/source/common/http/async_client_impl.cc#L180

the callback finishes executing, but we're still in `encodeHeaders`, which looks to see if `end_stream` is true, and if it is, goes ahead and calls `closeRemote`: https://github.com/envoyproxy/envoy/blob/ef543d8fcea4d1c0ce85e3ff8d16fb3928436ed6/source/common/http/async_client_impl.cc#L94

This is largely existing behavior, which this PR preserves, except for the added guards to prevent "extra" closures/resets.

Reply:
OK I see. That was my confusion that we are using `local_closed_` and `remote_closed_` also in the case of reset. IMO this is kind of confusing, so if you want to somehow split this logic out I would go for that. If not, can you potentially make the comments in these cases a lot more robust for the next reader to come along? I think that would be good enough for me. Thank's for explaining all of this.

/wait